File: modules/dev-menu-fps/android/src/main/java/expo/modules/devmenufps/DevMenuFpsModule.kt
// Expo module entry for Android. Bridges DevMenuFpsMonitor to JS via onFpsUpdate events.
package expo.modules.devmenufps

import android.content.Context
import android.os.Build
import android.view.Display
import android.view.WindowManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.math.roundToInt

class DevMenuFpsModule : Module() {
  private var monitor: DevMenuFpsMonitor? = null

  override fun definition() = ModuleDefinition {
    Name("DevMenuFps")

    Events("onFpsUpdate")

    Function("getMaxRefreshRate") {
      resolveMaxRefreshRate()
    }

    Function("startMonitoring") {
      val maxRefreshRate = resolveMaxRefreshRate()
      monitor?.stop()
      monitor = DevMenuFpsMonitor(maxRefreshRate) { uiFps ->
        sendEvent(
          "onFpsUpdate",
          mapOf(
            "uiFps" to uiFps,
          ),
        )
      }
      monitor?.start()
    }

    Function("stopMonitoring") {
      monitor?.stop()
      monitor = null
    }
  }

  private fun resolveMaxRefreshRate(): Int {
    val activity = appContext.currentActivity
    if (activity != null) {
      val display = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
        activity.display
      } else {
        @Suppress("DEPRECATION")
        activity.windowManager.defaultDisplay
      }
      return display?.refreshRate?.roundToInt()?.coerceAtLeast(60) ?: 60
    }

    val context = appContext.reactContext?.applicationContext ?: return 60
    val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
    val display: Display? =
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
        context.display
      } else {
        @Suppress("DEPRECATION")
        windowManager.defaultDisplay
      }

    return display?.refreshRate?.roundToInt()?.coerceAtLeast(60) ?: 60
  }
}

File: modules/dev-menu-fps/ios/DevMenuFpsModule.swift
// Expo module entry for iOS. Bridges DevMenuFpsMonitor to JS via onFpsUpdate events.
// Exposes start/stop monitoring and getMaxRefreshRate (UIScreen.maximumFramesPerSecond).
import ExpoModulesCore

private let onFpsUpdate = "onFpsUpdate"

public class DevMenuFpsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DevMenuFps")

    Events(onFpsUpdate)

    Function("startMonitoring") {
#if DEBUG
      DevMenuFpsMonitor.shared().start(onUpdate: { uiFps in
        self.sendEvent(onFpsUpdate, [
          "uiFps": uiFps,
        ])
      })
#endif
    }

    Function("stopMonitoring") {
      DevMenuFpsMonitor.shared().stop()
    }

    Function("getMaxRefreshRate") { () -> Int in
      Int(UIScreen.main.maximumFramesPerSecond)
    }
  }
}
